### PR TITLE
Components

### PR DESCRIPTION
**Pull request by lha047**
Add components for page elements

Relates to MT-2

### CODE REVIEW

**Review by tajakobsen — CHANGES_REQUESTED** [review]
Very good job! But there are a few things that I think should be looked at.

Note that some of the comments are very opinionated by me. So don't be afraid to challenge them if you disagree. And we'll decide on a common standard to write the code together.

Some of the main things that needs to be addressed are:
 1. I think some of the wrapping components are redundant
 2. Some components use ids that are not necessarily unique
 3. The components encapsulating input fields might be made more independent of functionality related to multipageform

Also see if it's possible to run prettier on the code.

File: src/lib/mapper/mapper.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+export const getSubmitButtonText = (pages: MultiPageFormPage[], currentPageTitle): string | null => {
+  const page = pages.find((page) => toSnakeCase(page.pageTitle) === currentPageTitle.toLowerCase());
+  return page?.submitButtonText ? page.submitButtonText : null;
+};
+
+export const mapErrorFromSchema = (error: SchemaError): UIError => {
+  return {
+    fieldName: findFieldNameForError(error),
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/lib/mapper/mapper.ts:76]: Will update :)

File: src/components/input/TextInput.svelte [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  let placeholder = fieldProps.properties.placeholder?.placeholder;
+</script>
+
+<Label labelText={label} labelFor={name} />
+<HelpText {helpText} />
+
+<InputError fieldName={name} {error} />
+
+<input
+  id={name}
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/input/TextInput.svelte:25]: Will make a task to add unique ids for the components we decide to keep

File: src/components/form/FormErrorSummary.svelte [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<script lang="ts">
+  import type { UIError } from '$lib/types/UIError';
+
+  export let errors: UIError[];
+  export let errorHeading: string;
+</script>
+
+{#if errors}
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/form/FormErrorSummary.svelte:8]: Will update check

File: src/components/form/InputErrorMessage.svelte [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  import type { UIError } from '$lib/types/UIError';
+  import ErrorIcon from '$components/icons/ErrorIcon.svelte';
+
+  export let error: UIError;
+</script>
+
+{#if error && error.fieldName}
+  <span id={`${error.fieldName}-error`} class="form-error">
+    <ErrorIcon />
+    <span class="inclusively-hidden">Feilmelding:</span>
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/form/InputErrorMessage.svelte:11]: Agree :)

File: src/lib/mapper/mapper.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
+import type { InputTypes, MultiPageFormPage, MultiPageFormPageWithSchema } from '$lib/types/MultipageForm';
+import type { ErrorObject } from 'ajv';
+import type { UIError } from '$lib/types/UIError';
+import type { SchemaError } from '$lib/types/JSONSchema';
+
+export const findProperties = (field: InputTypes) => {
+  const key = field._selected;
+  return field[key];
+};
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/lib/mapper/mapper.ts:9]: Just used to writing functions this way (from functions in React-components) 😂 
Will update :)

File: src/lib/mapper/mapper.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+export const findValue = (field: InputTypes, values) => {
+  const props = findProperties(field);
+  return values[props.name];
+};
+
+export const findError = (field: InputTypes, errors: UIError[]) => {
+  const props = findProperties(field);
+  return errors?.find(({ fieldName }) => fieldName === props.name);
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/lib/mapper/mapper.ts:18]: Willl fix

File: src/components/form/InputErrorMessage.svelte [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<script lang="ts">
+  import type { UIError } from '$lib/types/UIError';
+  import ErrorIcon from '$components/icons/ErrorIcon.svelte';
+
+  export let error: UIError;
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/form/InputErrorMessage.svelte:5]: Error passed down a few levels. My thought was to group the fields in an object would make it more consistent and less to pass down

File: src/components/input/RadioGroup.svelte [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  let selected = value ?? (setDefault && fieldProps?.options[0].name);
+</script>
+
+<fieldset id={name} aria-describedby={`${name}-hint ${name}-error`} class="form-fieldset">
+  <legend for={name} class="form-legend">{label}</legend>
+  <HelpText {helpText} />
+  <InputError {name} {error} />
+
+  {#each options as radio (radio.id)}
+    <div class="radio-item">
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/input/RadioGroup.svelte:31]: It was for styling. Woud you prefere to wrap the input with the label?

File: src/components/input/RadioGroup.svelte [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  {/if}
+
+  {#each options as radio (radio.name)}
+    <label class="form-label" for={radio.name}>
+      <input
+        type="radio"
+        id={radio.name}
+        {name}
+        bind:group={selected}
+        value={radio.name}
```
Comment (tajakobsen) [inline src/components/input/RadioGroup.svelte:34]: There should probably be a `radio.value` from XP too? Or is this enough?

File: src/components/input/RadioGroup.svelte [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      {@html helpText}
+    </div>
+  {/if}
+
+  {#if error}
+    <InputError {...error} />
+  {/if}
+
+  {#each options as radio (radio.name)}
+    <label class="form-label" for={radio.name}>
```
Comment (tajakobsen) [inline src/components/input/RadioGroup.svelte:28]: Stack the attributes.

File: src/components/input/TextArea.svelte [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+  export let value = undefined;
+  export let error: UIError | undefined = undefined;
+  export let name: string;
+  export let label: string;
+  export let helpText: string | undefined = undefined;
+  export let disclosure;
+</script>
+
+<label class="form-label" for={name}>
```
Comment (tajakobsen) [inline src/components/input/TextArea.svelte:13]: Stack attributes

File: src/components/input/TextArea.svelte [hunk 12]
```diff
@@ -0,0 +1,27 @@
+<script lang="ts">
+  import InputError from '$components/form/InputErrorMessage.svelte';
+  import { UIError } from '$lib/types/UIError';
+
+  export let value = undefined;
+  export let error: UIError | undefined = undefined;
+  export let name: string;
+  export let label: string;
+  export let helpText: string | undefined = undefined;
+  export let disclosure;
```
Comment (tajakobsen) [inline src/components/input/TextArea.svelte:10]: We don't need `disclosure` any more right?

File: src/components/input/TextArea.svelte [hunk 13]
```diff
@@ -0,0 +1,27 @@
+<script lang="ts">
+  import InputError from '$components/form/InputErrorMessage.svelte';
+  import { UIError } from '$lib/types/UIError';
+
+  export let value = undefined;
```
Comment (tajakobsen) [inline src/components/input/TextArea.svelte:5]: We probably don't need to initialize variables to `undefined`

File: src/components/input/TextArea.svelte [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  <div class="hint">
+    {@html helpText}
+  </div>
+{/if}
+
+{#if error}
+  <InputError {...error} />
+{/if}
+
+<textarea id={name} {name} bind:value aria-describedby={`${name}-hint ${name}-error`} aria-invalid={!!error.message} />
```
Comment (tajakobsen) [inline src/components/input/TextArea.svelte:27]: Stack attributes vertifally

File: src/components/input/TextInput.svelte [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  import InputError from '$components/form/InputErrorMessage.svelte';
+  import type { UIError } from '$lib/types/UIError';
+
+  export let value = undefined;
+  export let error: UIError | undefined = undefined;
+  export let name: string;
+  export let label: string;
+  export let helpText: string | undefined = undefined;
+  export let properties;
+  export let disclosure;
```
Comment (tajakobsen) [inline src/components/input/TextInput.svelte:11]: `disclosure` is not in use

File: src/components/input/TextInput.svelte [hunk 16]
```diff
@@ -0,0 +1,42 @@
+<script lang="ts">
+  import InputError from '$components/form/InputErrorMessage.svelte';
+  import type { UIError } from '$lib/types/UIError';
+
+  export let value = undefined;
```
Comment (tajakobsen) [inline src/components/input/TextInput.svelte:5]: Do we need to default to `undefined`?

File: src/components/input/TextInput.svelte [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  export let label: string;
+  export let helpText: string | undefined = undefined;
+  export let properties;
+  export let disclosure;
+  let inputMode = properties.inputMode?.inputmode;
+  let autocomplete = properties.autocomplete?.autocomplete;
+  let placeholder = properties.placeholder?.placeholder;
+</script>
+
+<label class="form-label" for={name}>
```
Comment (tajakobsen) [inline src/components/input/TextInput.svelte:17]: Stack attributes

File: src/components/input/TextInput.svelte [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+  {name}
+  class="form-field"
+  bind:value
+  class:error
+  aria-describedby={`${name}-hint ${name}-error`}
+  aria-invalid={!!error}
+  inputmode={inputMode}
+  {placeholder}
+  {autocomplete}
+/>
```
Comment (tajakobsen) [inline src/components/input/TextInput.svelte:42]: Move `/>` to line above

File: src/components/input/TextInput.test.ts [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+      inputmode: 'text',
+    },
+    autocomplete: {
+      autocomplete: 'name',
+    },
+    placeholder: {
+      placeholder: 'Eg. Test Testsen',
+    },
+  };
+  let disclosure;
```
Comment (tajakobsen) [inline src/components/input/TextInput.test.ts:21]: Is `disclosure` used in the components anymore?

**Review by tajakobsen — CHANGES_REQUESTED** [review]
Good job! There are mainly code standard suggestions here now. Should be quick to fix.

File: src/components/form/FormErrorSummary.svelte [hunk 20]
```diff
@@ -0,0 +1,19 @@
+<script lang="ts">
+  import type { UIError } from '$lib/types/UIError';
+
+  export let errors: UIError[];
+  export let heading: string;
+</script>
+
+{#if errors.length > 0}
+  <div class="error-summary" role="alert" tabindex="-1" aria-labelledby="error-summary-heading">
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/components/form/FormErrorSummary.svelte:9]: Automatic format

**Review by tajakobsen — APPROVED** [review]
Very good job! Closing and merging to *main*.

**Pull request merged**